demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb

```diff
@@ -0,0 +1,631 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 54,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "llm = HuggingFaceTextGenInference(\n",
+    "    inference_server_url=LLAMA2_7B_HOSTPORT,\n",
+    "    max_new_tokens=512,\n",
+    "    top_k=10,\n",
+    "    top_p=0.9,\n",
+    "    typical_p=0.95,\n",
+    "    temperature=0.6,\n",
+    "    repetition_penalty=1,\n",
+    "    do_sample=True,\n",
+    "    streaming=True\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next, we define the retriever and template for our RetrivalQA chain. For each call of the RetrievalQA, LangChain performs a semantic similarity search of the query in the vector database, then passes the search results as the context to Llama to answer the query about the data stored in the verctor database.  \n",
+    "Whereas for the template, this defines the format of the question along with context that we will be sent into modes for generation. In general, Llama 2 has special prompt format to handle special tokens. In some cases, the serving framework might already have taken care of it. Otherwise, you will need to write customized template to properly handle that.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 81,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "template = \"\"\"\n",
+    "[INST]Use the following pieces of context to answer the question.\n",
+    "{context}\n",
+    "Question: {question} [/INST]\n",
+    "\"\"\"\n",
+    "\n",
+    "retriever = db.as_retriever(\n",
+    "        search_kwargs={\"k\": 6}\n",
+    "    )"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Lastly, we can define the retrieval chain for QA"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 82,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "qa_chain = RetrievalQA.from_chain_type(\n",
+    "    llm=llm, \n",
+    "    retriever=retriever,     \n",
+    "    chain_type_kwargs={\n",
+    "        \"prompt\": PromptTemplate(\n",
+    "            template=template,\n",
+    "            input_variables=[\"context\", \"question\"],\n",
+    "        ),\n",
+    "    }\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we should have a working chain for QA. Let's test it out before wire it up with UI blocks."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "result = qa_chain({\"query\": \"Why choose Llama?\"})\n",
+    "print(result)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "After confirming the validity, we can start building the UI. Before we define the gradio blocks, let's first define the callback streams that we will use later for streaming feature.  \n",
```

Comment:
A quick explanation of the gradio blocks?

streaming feature -> the streaming feature

Reply:
Added the documentation link. Better explanation there.